granary/reddit.py

```diff
@@ -0,0 +1,218 @@
+import praw
+
+if appengine_info.DEBUG:
+  REDDIT_APP_KEY = util.read('reddit_app_key_local')
+  REDDIT_APP_SECRET = util.read('reddit_app_secret_local')
+else:
+  REDDIT_APP_KEY = util.read('reddit_app_key')
+  REDDIT_APP_SECRET = util.read('reddit_app_secret')
+
+
+def get_reddit_api(refresh_token):
+  return praw.Reddit(client_id=REDDIT_APP_KEY,
+             client_secret=REDDIT_APP_SECRET,
+             refresh_token=refresh_token,
+             user_agent='oauth-dropin reddit api')
+
+
+class Reddit(source.Source):
+
+  DOMAIN = 'reddit.com'
+  BASE_URL = 'https://reddit.com'
+  NAME = 'reddit'
+
+  def __init__(self, refresh_token):
+    self.refresh_token = refresh_token
+
+  def user_to_actor(self, user):
+    """Converts a praw Redditor to an actor.
+    https://praw.readthedocs.io/en/latest/code_overview/models/redditor.html
+
+    Args:
+      user: praw Redditor object
+
+    Returns:
+      an ActivityStreams actor dict, ready to be JSON-encoded
+    """
+    username = user.name
+    if not username:
+      return {}
+
+
+    # trying my best to grab all the urls from the profile description
+
+    description = ''
+    if user.subreddit:
+      user_url = self.BASE_URL + user.subreddit.get('url')
+      urls = [user_url]
+      description = user.subreddit['public_description']
+      profile_urls = util.extract_links(description)
+      urls += util.trim_nulls(profile_urls)
+    else:
+      urls = [self.BASE_URL + '/user/' + username]
+
+    image = user.icon_img
+
+    return util.trim_nulls({
+      'objectType': 'person',
+      'displayName': username,
+      'image': {'url': image},
+      'id': self.tag_uri(username),
+      # numeric_id is our own custom field that always has the source's numeric
+      # user id, if available.
+      'numeric_id': user.id,
+      'published': util.maybe_timestamp_to_iso8601(user.created_utc),
+      'url': urls[0],
+      'urls': [{'value': u} for u in urls] if len(urls) > 1 else None,
+      'username': username,
+      'description': description,
+      })
+
+
+  def praw_to_object(self, thing, type):
+    """
+    Converts a praw object to an object.
+
+    Args:
+      thng: a praw object, Sumbission or Comment
+      type: string to denote whether to get submission or comment content
+
+    Returns:
+      an ActivityStreams object dict, ready to be JSON-encoded
+      """
+    obj = {}
+
+    # always prefer id_str over id to avoid any chance of integer overflow.
+    # usually shouldn't matter in Python, but still.
+    id = thing.id
+    if not id:
+      return {}
+
+    published = util.maybe_timestamp_to_iso8601(thing.created_utc)
+
+    obj = {
+      'id': self.tag_uri(id),
+      'objectType': 'note',
+      'published': published,
+      'to': [],
+      }
+
+    user = thing.author
+    if user:
+      obj['author'] = self.user_to_actor(user)
+      username = obj['author'].get('username')
+
+    obj['to'].append({
+      'objectType': 'group',
+      'alias': '@public',
```

Comment:
all reddit content is public? if so probably just inline this into line 114?

Reply:
i think there are non-public things, but I don't plan to query them at least initially.  i've made a note in the docstring. and maybe I can file an issue to double back to handle private things in the future?

Reply:
sgtm! as long as you know for sure that you'll only ever get public things from your queries...?